File: rust/src/implementations/tetra/display.rs
use crate::chip8::{Output, COLS, ROWS, SCALE};
use tetra::graphics::{self, Canvas, Color, DrawParams, Texture};
use tetra::math::Vec2;
use tetra::Context;

pub struct Display {
    canvas: Canvas,
    texture: Texture,
    pub draw_flag: bool,
    pub frame_buffer: [bool; COLS as usize * ROWS as usize], // display buffer
}

impl Display {
    pub fn new(ctx: &mut Context) -> tetra::Result<Self> {
        Ok(Display {
            canvas: Canvas::new(ctx, COLS as i32, ROWS as i32)?,
            texture: Texture::new(ctx, "./resources/textures/square.png")?,
            draw_flag: false,
            frame_buffer: [false; COLS as usize * ROWS as usize],
        })
    }

    pub fn draw(&mut self, ctx: &mut Context) {
        if self.draw_flag {
            graphics::set_canvas(ctx, &self.canvas);
            graphics::clear(ctx, Color::BLACK);
            for y in 0..ROWS {
                for x in 0..COLS {
                    if self.frame_buffer[y as usize * COLS as usize + x as usize] {
                        self.texture.draw(
                            ctx,
                            DrawParams::new().position(Vec2::new(x as f32, y as f32)),
                        );
                    }
                }
            }
            graphics::reset_canvas(ctx);
            self.draw_flag = false;
        }
        self.canvas.draw(
            ctx,
            DrawParams::new()
                .position(Vec2::zero())
                .scale(Vec2 { x: SCALE, y: SCALE }),
        );
    }
}

impl Output for Display {
    fn clear(&mut self) {
        self.frame_buffer = [false; 64 * 32];
        self.draw_flag = true;
    }

    fn get_pixel(&self, addr: usize) -> bool {
        self.frame_buffer[addr]
    }

    fn set_pixel(&mut self, addr: usize, value: bool) {
        self.frame_buffer[addr] = value;
    }

    fn flip_pixel(&mut self, addr: usize) {
        self.frame_buffer[addr] = !self.frame_buffer[addr];
    }

    fn set_draw_flag(&mut self, value: bool) {
        self.draw_flag = value;
    }
}

File: typescript/src/display.ts
interface IDisplay {
    clear() : void;
    get Width() : number;
    get Height() : number;
    getPixel(xCoord : number, yCoord : number) : boolean;
    setPixel(xCoord : number, yCoord : number) : void;
    setDrawFlag() : void;

    draw(ctx : CanvasRenderingContext2D) : void;
}

class Display implements IDisplay {
    _pixels : boolean[];
    _width : number;
    _height : number;
    _drawFlag : boolean;

    constructor(width : number, height : number) {
        this._width = width;
        this._height = height;
        this._pixels = new Array(width * height).fill(false);
    }

    get Width(): number {
        return this._width;
    }

    get Height(): number {
        return this._height;
    }
    
    getPixel(xCoord: number, yCoord: number): boolean {
        return this._pixels[(yCoord * this.Width) + xCoord];
    }
    
    setPixel(xCoord: number, yCoord: number): void {
        this._pixels[(yCoord * this.Width) + xCoord] = !this._pixels[(yCoord * this.Width) + xCoord];
    }

    clear() : void {
        this._pixels = new Array(this.Width * this.Height).fill(false);
    }

    setDrawFlag(): void {
        this._drawFlag = true;
    }

    draw(ctx : CanvasRenderingContext2D) : void {
        if(this._drawFlag) {
            ctx.fillStyle = "black";
            ctx.clearRect(0, 0, this.Width * 10, this.Height * 10);
            ctx.fillStyle = "black";
            for (let index = 0; index < this._pixels.length; index++) {
                if(this._pixels[index]) {
                    let x = (index % this.Width) * 10;
                    let y = Math.floor(index / this.Width) * 10;
                    ctx.fillRect(x, y, 10, 10);
                }
            }
            this._drawFlag = false;
        }
    }
}
